### ISSUE TITLE
SSL_get_error returns a confused error in async mode

### ISSUE THREAD

**Issue by yangyangtiantianlonglong** [openssl/openssl#16809]
In async mode（`SSL_CTX_set_mode(ssl, SSL_MODE_ASYNC)`）, When `SSL_read` receives the `SSL_shutdown` message from the peer, `SSL_get_error` returns `SSL_ERROR_WANT_ASYNC`.  However, ASYNC_start_job returns `ASYNC_FINISH`.

In normal mode，When `SSL_read` receives the `SSL_shutdown` message from the peer, `SSL_get_error` returns `SSL_ERROR_ZERO_RETURN`.

**Comment by yangyangtiantianlonglong** [comment]
https://github.com/openssl/openssl/blob/58608487a44b3991ecc6d431d6273b2ca8c980a6/ssl/ssl_lib.c#L1760-L1794
I guess `s->rwstate` is not refreshed, and `ASYNC_FINISH` is returned in `ASYNC_start_job`.

**Comment by kaduk** [comment]
I would need to look more closely to be sure, but setting `s->rwstate = SSL_NOTHING` just because we received a close_notify alert seems incompatible with us sending a "close_notify" in response, as the RFC encourages us to do.

**Comment by yangyangtiantianlonglong** [comment]
>  I think probably the issue is failing to reset the rwstate when we've finished. 

@mattcaswell   yes，You are right.


> Does this solve the problem:

This may cause a new problem ,  `s->rwstate`  set to `SSL_READING` when the `SSL_read` have finished in non-blocking, we will 'not get `SSL_ERROR_WANT_READ`.

**Comment by yangyangtiantianlonglong** [comment]
> I would need to look more closely to be sure, but setting `s->rwstate = SSL_NOTHING` just because we received a close_notify alert seems incompatible with us sending a "close_notify" in response, as the RFC encourages us to do.

@kaduk  I didn't understand the specific scenario you were talking about, `s->rwstate = SSL_NOTHING` does not affect the application to respond with the "close_notify".

**Comment by yangyangtiantianlonglong** [comment]
@mattcaswell  
I didn't think of a better way,Do you have a better idea?   The #16820  can solve the problem.

**Comment by mattcaswell** [comment]
> The s->rwstate cannot be restarted. The s->rwstate is used when the I/O function is ASYNC_FINISH.
For example, s->rwstate = SSL_READING when the SSL_read had ASYNC_FINISH in non-blocking, if s->rwstate = SSL_NOTHING, we will 'not get SSL_ERROR_WANT_READ.

Good point.

> I didn't think of a better way,Do you have a better idea? 

How about resetting the rwstate *before* the `ASYNC_start_job` call, i.e.

```` diff
diff --git a/ssl/ssl_lib.c b/ssl/ssl_lib.c
index 3d20305170..a37eafca67 100644
--- a/ssl/ssl_lib.c
+++ b/ssl/ssl_lib.c
@@ -1770,6 +1770,8 @@ static int ssl_start_async_job(SSL *s, struct ssl_async_args *args,
                  (s->waitctx, ssl_async_wait_ctx_cb, s))
             return -1;
     }
+    if (s->rwstate == SSL_ASYNC_PAUSED)
+        s->rwstate = SSL_NOTHING;
     switch (ASYNC_start_job(&s->job, s->waitctx, &ret, func, args,
                             sizeof(struct ssl_async_args))) {
     case ASYNC_ERR:

````

**Comment by yangyangtiantianlonglong** [comment]
> How about resetting the rwstate before the ASYNC_start_job call, i.e.

yes, It's always wise to do reset before the ASYNC_start_job call, i.e.

```javascript
diff --git a/ssl/ssl_lib.c b/ssl/ssl_lib.c
index 3d20305170..a37eafca67 100644
--- a/ssl/ssl_lib.c
+++ b/ssl/ssl_lib.c
@@ -1770,6 +1770,8 @@ static int ssl_start_async_job(SSL *s, struct ssl_async_args *args,
                  (s->waitctx, ssl_async_wait_ctx_cb, s))
             return -1;
     }

+    s->rwstate = SSL_NOTHING;
     switch (ASYNC_start_job(&s->job, s->waitctx, &ret, func, args,
                             sizeof(struct ssl_async_args))) {
     case ASYNC_ERR:
```
 The `s->rwstate == SSL_ASYNC_PAUSED` is unnecessary, because the `s->rwstate`  can be updated When other functions are complete. In openssl, only `SSL_get_error()` checks the `s->rwstate` , and `SSL_get_error()` is always be called after other functions are complete.

**Comment by mattcaswell** [comment]
Fix for this in #17013
